### ISSUE TITLE
BadHttpRequestException: Reading the request body timed out due to data arriving too slowly

### ISSUE THREAD

**Issue by clement911** [dotnet/aspnetcore#4707]
We have an asp.net core 2.1 app running on .net 4.7 in azure web app service.

Recently we started getting A LOT of the following error:

> Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException: Reading the request body timed out due to data arriving too slowly. See MinRequestBodyDataRate.

Our app is under pretty constant heavy load (from incoming webhooks mostly). During big spikes, we tend to see more of these.

We used to run on asp.net core 2.0 / .net 4.6.1 for many months and we never saw that error before.
It seems to have started happening following our recent upgrade to asp.net core 2.1 / .net 4.7.

We'd like to get to the bottom of this issue, but we are not even sure where to start.
Any thoughts?

**Comment by chadgruka** [comment]
We're seeing the same error messages, but we were able to reproduce them using Chrome's Network Throttling profiles. After creating one at 1kb/s (up/down) and 5K ms latency, we were able to see the effects of getting BadHttpRequestExceptions with "Reading the request body timed out due to data arriving too slowly. See MinRequestBodyDataRate".

In our production case the issue was a mobile client, with very spotty internet connection, downloading data.

@davidfowl Is a hard exception the best way to deal with this? Do we just need middleware to handle slow clients, and swallow the exception?

As in our case, the app services are being recycled (assumed due to something like rapid fail protection). I don't recall anything like this in IIS 7 running ASP.NET 4.5. Although, IIS would have disabled the app pool, and taking the application offline completely...

**Comment by highfield** [comment]
The same problem to me as well.
Here is my code serving photo uploads from several cameras connected via GPRS (powering on then off to save energy):

```
        [HttpPost("photo/post")]
        public async Task PostLegacyAsync()
        {
            var content = new MultipartFormDataContent();
            foreach (var pair in this.Request.Form)
            {
                content.Add(new StringContent(pair.Value), pair.Key);
            }
            foreach (IFormFile ff in this.Request.Form.Files)
            {
                byte[] data;
                using (var br = new BinaryReader(ff.OpenReadStream()))
                {
                    data = br.ReadBytes((int)ff.OpenReadStream().Length);
                }
                content.Add(new ByteArrayContent(data), ff.Name, ff.FileName);
            }

            // ...

            this.Ok();
        }

```

This service is considered legacy for our system, but there are still several devices to upgrade. Suggestions to improve it are welcome.

As a side note, I could accept a random failure sometime, but I don't like the application crashing for that.

**Comment by davidfowl** [comment]
@highfield you should avoid using Request.Form and instead use ReadFormAsync

**Comment by Max-Eriksson** [comment]
Hi All,

We have the same issue in production. In our environment it seems to be related to bursts in connections/requests, see attached diagram.


![image](https://user-images.githubusercontent.com/10500964/51591276-e08d9100-1eec-11e9-810f-04a33ebb9c9a.png)

**Comment by clement911** [comment]
This is just to say that we are still very much struggling with this issue.
I'm hoping future releases will be able to handle spikes better.

**Comment by alastairs** [comment]
We're seeing the same issue in production, along with a related `BadHttpRequestException` triggering a 400 response: "Unexpected end of request content". We have little insight into what is causing this problem, but as little as 1 request/second with a 600KB request body will trigger it. It seems to be exacerbated by unreliable client connections, such as over mobile/cellular 3G networks. We see no related spikes in CPU or memory consumption at the time when these occur.

**Comment by BrandonPotter** [comment]
Have been trying to track down this issue as to why ASP.NET Core 2.1 containers keep crashing on K8s. Thought it would abort the request but surprised that it seems to crash the process...

**Comment by davidfowl** [comment]
Aborting the request doesn't crash the process

**Comment by Chakalaka11** [comment]
> This is just to say that we are still very much struggling with this issue.
> I'm hoping future releases will be able to handle spikes better.

@clement911 i was able to fix this issue in my work project by removing usage of HTTP2 in Kestrel configuration in Program.cs. Could you try and reply about effect - I really curious about it because in company only out project has this issue.

**Comment by clement911** [comment]
@Chakalaka11 we have not enabled http 2 but could you please copy your code here?

**Comment by Chakalaka11** [comment]
@clement911 So, for the past hours team made some changes, removing UseKestrel, using ConfigureKestrel to re-attach http2, work also fine. Guy who made fix says that UseKestrel could rewrite some standard config and made error.
![image](https://user-images.githubusercontent.com/12372021/54744985-1b503480-4bd1-11e9-805d-4e599c46f206.png)

**Comment by Chakalaka11** [comment]
Maybe this help 
Today i will make some regression tests to check how all works and update status if it works fine.

**Comment by clement911** [comment]
Thanks but I'm hesitant to apply changes that don't seem to be related to the original issue.
We'd like to use in-process host but I see it's not available for .net framework...

**Comment by Chakalaka11** [comment]
Ok, if I will have some updates on this error - I'll let you know. Hope you will find your solution.

**Comment by ttutisani** [comment]
I'd like to join the "party".
In our code, we do:
```
await context.Request.Body.CopyToAsync(requestBodyStream);
```
Nothing synchronous aginst the request stream. We get the error Reading the request body timed out due to data arriving too slowly. See MinRequestBodyDataRate.

Any findings that could help us? I read the thread above and it sounds like the root cause is unknown so far.

**Comment by ttutisani** [comment]
So, I'm looking at [this](https://github.com/aspnet/KestrelHttpServer/issues/2855). Has anybody tried setting a lower value on `MinRequestBodyDataRate`?
Also, our application deals with slow network connections, indeed (by design). So maybe we should instruct the front end to retry when the connection is improved? Again, this may apply to only certain kinds of apps, and perhaps we are among them.

**Comment by BrandonPotter** [comment]
I did try the following in `Startup.cs` `ConfigureServices` method...

```
services.Configure<KestrelServerOptions>(options =>
            {
                options.Limits.MinRequestBodyDataRate = new MinDataRate(1, TimeSpan.FromMinutes(5));
                options.Limits.MinResponseDataRate = new MinDataRate(1, TimeSpan.FromMinutes(5));
                options.Limits.MaxRequestBodySize = 100000000;
            });
```

Prior to making this change, I had asp.net core containers in Kubernetes that were crashing/restarting 90x an hour, with the last message in the container log being the `BadHttpRequestException` stack trace. (As @davidfowl said above, this seemed really odd because I would expect this exception to occur at request level and not crash the process... so I'm inclined to think that perhaps there was something else also contributing to the crash?)

After making this change, have not had this issue. If this issue continued, my plan was to try and put some sort of reverse proxy in front of the Kestrel server that would buffer the slow incoming data stream, but haven't had to go to that point yet.

**Comment by clement911** [comment]
Good to hear that some are finding solutions.
Personally, I don't like to apply work arounds I don't really understand without understanding the issue first.

As far as I understand, under heavy load, the "Reading the request body timed out due to data arriving too slowly" error is not the real cause. It is a consequence of the system being stressed.
What I still haven't figured out is WHY the system is stressed given that metrics look healthy (CPU, memory and thread countmostly).
I'm still suspecting that it might be caused by the ThreadPool algorithm not creating threads fast enough when a spike of requests occurs, but as far as I know we cannot tweak this algorithm.

It would be fantastic if the asp.net core team could create a spike stress test on a simple web app to confirm if there is indeed a bottleneck at the thread pool level, or somewhere else.

**Comment by clement911** [comment]
We've had to upscale our azure instances even though they look healthy, just because we get tons of these, as well as 'Connection forcibly closed' and 'Connection processing ended abnormally' errors.

**Comment by guylando** [comment]
This happens for us in 
`Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1MessageBody.<PumpAsync>`
called by
`Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.MessageBody.<ReadAsync>`
called by 
`Microsoft.AspNetCore.WebUtilities.FormReader.<ReadFormAsync>`
called by
`Microsoft.AspNetCore.Antiforgery.Internal.DefaultAntiforgeryTokenStore.<GetRequestTokensAsync>`
which means that it happens at csrf check.

More stack trace:
`
Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException: Reading the request body timed out due to data arriving too slowly. See MinRequestBodyDataRate. at Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException.Throw(RequestRejectionReason reason) at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1MessageBody.<PumpAsync>d__4.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.IO.Pipelines.PipeCompletion.ThrowLatchedException() at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result) at System.IO.Pipelines.Pipe.GetReadAsyncResult() at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.MessageBody.<ReadAsync>d__21.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.<ReadAsyncInternal>d__21.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.IO.StreamReader.<ReadBufferAsync>d__97.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at System.IO.StreamReader.<ReadAsyncInternal>d__64.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at Microsoft.AspNetCore.WebUtilities.FormReader.<BufferAsync>d__41.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at Microsoft.AspNetCore.WebUtilities.FormReader.<ReadNextPairAsyncImpl>d__34.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at Microsoft.AspNetCore.WebUtilities.FormReader.<ReadFormAsync>d__43.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at Microsoft.AspNetCore.Http.Features.FormFeature.<InnerReadFormAsync>d__18.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at Microsoft.AspNetCore.Antiforgery.Internal.DefaultAntiforgeryTokenStore.<GetRequestTokensAsync>d__3.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at Microsoft.AspNetCore.Antiforgery.Internal.DefaultAntiforgery.<ValidateRequestAsync>d__9.MoveNext() --- End of stack trace from previous location where exception was thrown --- at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw() at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task) at Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.ValidateAntiforgeryTokenAuthorizationFilter.<OnAuthorizationAsync>d__3.MoveNext() -
`

**Comment by vankampenp** [comment]
We had this once now. The operation sends a ID string and an answer string, together with some cookies. so really small body. 

```
2019-05-09 11:31:50.9573|30131|TRACE|9.5.3.0|189|Base|BaseController.QuestionsSaveAnswer|27 ms|
2019-05-09 11:32:17.6812|30131|TRACE|9.5.3.0|97 |Base|BaseController.QuestionsSaveAnswer|24 ms|
2019-05-09 11:32:27.2114|30131|ERROR|9.5.3.0|106|Base|BaseController.GetErrorMessage|Error Page Path=/questionsSaveAnswer, error = Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException: Reading the request body timed out due to data arriving too slowly. See MinRequestBodyDataRate.
   at Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException.Throw(RequestRejectionReason reason)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1MessageBody.PumpAsync()
   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.GetReadAsyncResult()
   at System.IO.Pipelines.Pipe.DefaultPipeReader.GetResult(Int16 token)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.MessageBody.ReadAsync(Memory`1 buffer, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.WebUtilities.FileBufferingReadStream.ReadAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.WebUtilities.StreamHelperExtensions.DrainAsync(Stream stream, ArrayPool`1 bytePool, Nullable`1 limit, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Mvc.Formatters.JsonInputFormatter.ReadRequestBodyAsync(InputFormatterContext context, Encoding encoding)
   at Microsoft.AspNetCore.Mvc.ModelBinding.Binders.BodyModelBinder.BindModelAsync(ModelBindingContext bindingContext)
   at Microsoft.AspNetCore.Mvc.ModelBinding.ParameterBinder.BindModelAsync(ActionContext actionContext, IModelBinder modelBinder, IValueProvider valueProvider, ParameterDescriptor parameter, ModelMetadata metadata, Object value)
   at Microsoft.AspNetCore.Mvc.Internal.ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<<CreateBinderDelegate>g__Bind|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
```
 We are running in-process ASP Net Core 2.2.0 on IIS. In the log the number before |Base| is the Thread ID.

**Comment by clement911** [comment]
We finally found a fix!
As we had suspected, it had to do with the thread pool not ramping up threads quickly enough.
We added the following two lines and this fixed the problem entirely and immediately:

```
ThreadPool.SetMaxThreads(32_767, 2_000);
ThreadPool.SetMinThreads(1_000, 1_000);
```

The thread pool ramp up algorithm is very conservative, and so are the default minimum threads.
I'm not actually sure if we needed to ramp up the worker threads. Possibly the io threads would be enough, as the CPU at a very low percentage utilization.
So my reading of this is that the IO requests were being queued up internally by the thread pool under spike.
I'm glad we found a solution, but even though the fix is tiny, it took a while to find it.

IMO the default are too conservative.
For a regular website that might work fine, but for an API or webhook receiver, there will be spikes that need to ramp up much faster than 2 threads per second.

It doesn't help that the ThreadPool is opaque and the documentation is very limited.

**Comment by davidfowl** [comment]
cc @stephentoub @kouvel

**Comment by vankampenp** [comment]
This is an excellent article describing what is causing this and why interesting the minimum should only be a stopgap measure
https://blogs.msdn.microsoft.com/vancem/2018/10/16/diagnosing-net-core-threadpool-starvation-with-perfview-why-my-service-is-not-saturating-all-cores-or-seems-to-stall/

**Comment by clement911** [comment]
Yes thank you, I had read that one and it is indeed a great article.
It says that increasing the minimum of threads is a stopgap measure **if the thread starvation is caused by blocking calls**.
But in our case, we don't ever block. We only use async IO APIs.
So, as far as I can tell, the only way to handle sudden burst is to increase the minimum of threads.
We have now scaled down to a cheaper azure instance and scaled in from 4 instances to 2 instances and we are not getting a single error anymore. It's pretty remarkable!
Interestingly, the number of threads has not increased substantially but it can just ramp up **much** quicker than before.
Also note that the number of threads is below the minimum of 1,000 that we set, because the thread pool doesn't spawn them unless there are items waiting in the queue.
At least, that's my understanding.
It might be useful to be able to instantiate thread pools and give them different options for different workloads...

**Comment by davidfowl** [comment]
I'm *extremely* surprised that you're seeing starvation without blocking, even for bursts.

**Comment by guylando** [comment]
@clement911 maybe try to use https://github.com/benaadams/Ben.BlockingDetector to see which framework code is blocking

**Comment by kouvel** [comment]
The only things that I'm aware of that would cause thread starvation are blocking, or long-running CPU/memory-bound work items in the thread pool. @clement911 since you mentioned that CPU utilization is low, without more information I would guess that blocking is occurring.

Blocking doesn't necessarily have to occur from synchronous IO. It could be a lock that is held for longer than expected, forward dependencies in queued tasks (a task schedules another task and synchronously waits for it to complete), or generally any synchronous activity that takes some time.

> We finally found a fix!
> As we had suspected, it had to do with the thread pool not ramping up threads quickly enough.
> We added the following two lines and this fixed the problem entirely and immediately

That is the suggested workaround when it's not possible to fix the root cause. It may well be an appropriate solution if blocking is occurring, due to deficiencies in the thread pool when it comes to work items that block for relatively long spans of time. The solution may be counterproductive if blocking is not occurring. Again since that's not what you're seeing it sounds like blocking is occurring.

It may help to look at blocking that is likely occurring with a [PerfView](https://github.com/microsoft/perfview/blob/master/documentation/Downloading.md) trace including `Thread Time` (which includes context-switch events and time spent blocking).

**Comment by clement911** [comment]
Thanks I appreciate the comments.

A while back we didn't know what to do with this issue but we knew it was related to our webhook receiver controller action receiving a lot of requests.
So we deployed another separate azure app service instance, specifically to receive webhooks.
So, what we have is an asp.net core service with a single controller and a single action being called.
(We left the rest of the code, but the only action being called is the webhook one and it doesn't serve user requests or run any jobs.)
All it does it validate the headers to make sure the request is authentic and save the payload to a sql azure database with an async call.
That's it.
That's why I'm quite confident we don't have any blocking going on.

To your point, sure async improves scalability but there is still a limit to how many requests can be processed, right?
In our case, our spike can be very sudden and very large.
If a user initiates a CSV import in Shopify, Shopify will send us 1 request per row instantly.
We could get 1,000 requests hitting a single instance in a minute. Isn't it possible that under these conditions the thread pool will start queueing requests?

**Comment by davidfowl** [comment]
When running on IIS there’s a request queue deep in the kernel (http.sys) but there’s no request queuing happening in ASP.NET core. The only queue is the thread pool queue. The exception you’re seeing is coming from kestrel (so I’m assuming you’re using the out of process IIS module). If you can still reproduce the issue, a trace would be great. We’re just adding counters in .NET Core 3.0 which is of no help to you right now but this sorta thing should be more obvious in the future with those counters.

**Comment by clement911** [comment]
btw, I was referring to the thread pool queue.

**Comment by vankampenp** [comment]
Perhaps you could put the requests from Shopify in your own background  queue and process them one by one. That way, you never have the spike.
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/hosted-services?view=aspnetcore-2.2

I have done that in a similar scenario where the user uploads a spreadsheet and then lots of work has to be done per row. Happy to share the code somewhere.

**Comment by clement911** [comment]
@vankampenp 
The issue is that we must be sure that the webhook requests are persisted before we can respond HTTP 200 to Shopify.
If we just put it in a memory queue and the app restarts for any reason, then we would lose the requests.

**Comment by vankampenp** [comment]
@clement911 I don't know your particular situation. In our scenario we persist a key, then push it on the queue, which sets a flag once it is done. We run a daily catch all to see if there are any key's without flag. 
This only makes sense if the work that has to be done takes a long time, in our case about 1s per row.

**Comment by clement911** [comment]
even if it is just a key, the point is you need to persist at least something so IO is needed.

**Comment by berndku** [comment]
We had the same issue and solved it by moving the reading of the request body out of the controller. 

In our case we had a high volume REST API where first the request body was read inside of the controller with _HttpRequestStreamReader_. Under heavier load we saw every few days this error:

> BadHttpRequestException: Reading the request body timed out due to data arriving too slowly

And also faced a heavy drop of throughput during this time. This occured randomly and recovered by itself after about 10 minutes. 

With the BlockingDetector tool we were able to identify some minimal blocking somewhere in the framework while reading the request body that way.

Because of that we created an InputFormatter  that allowed to read json-formatted request body as string and retrieved the request body as parameter in the controller like that:

`public async Task<IActionResult> ProcessAsync([FromBody]string jsonBody = null)`

After applying this workaround, the BlockingDetector did not detect blocking anymore and we had no more BadHttpRequestExceptions for three weeks now.

**Comment by davidfowl** [comment]
> With the BlockingDetector tool we were able to identify some minimal blocking somewhere in the framework while reading the request body that way.

You can share the results of this?

**Comment by berndku** [comment]
Yes, that was the output of the tool: 
![image](https://user-images.githubusercontent.com/4060485/64549347-d65fbc00-d330-11e9-8376-d66f323481c2.png)

And here the previous controller code that triggered this: 
`using (HttpRequestStreamReader reader = new HttpRequestStreamReader(Request.Body, Encoding.UTF8))
                {
                    _jsonBody = await reader.ReadToEndAsync();
                }`

**Comment by davidfowl** [comment]
> using (HttpRequestStreamReader reader = new HttpRequestStreamReader(Request.Body, Encoding.UTF8)) { _jsonBody = await reader.ReadToEndAsync(); }

Was this code you wrote in your controller?

**Comment by berndku** [comment]
Yes, this was the code with the blocking.

**Comment by davidfowl** [comment]
@berndku why do manual JSON de-serialization from a string instead of using the built in formatter?

**Comment by berndku** [comment]
@davidfowl: performance, not all of our requests require complete de-serialization of the entire body. But I believe we also tested it quickly with the built in formatter by replacing the string with a JObject parameter and didn´t see blocking there as well.

**Comment by davidfowl** [comment]
> @davidfowl: performance, not all of our requests require complete de-serialization of the entire body. But I believe we also tested it quickly with the built in formatter by replacing the string with a JObject parameter and didn´t see blocking there as well.

Performance is overloaded? Better throughput trading off memory usage? Buffering the entire body into memory means you have this temporary string just for the purpose of turning that into another object. The built in formatters don't block, they go out of their way to avoid buffering into a single contiguous block of memory (like a string or byte[]). 

Regardless, I'll file an issue for the ReadToEndAsync behavior https://github.com/aspnet/AspNetCore/issues/13834

**Comment by berndku** [comment]
Thanks for filing the issue!

Fully understand your point about built-in formatters and buffering, but we have to trace each body completely within the controller, no matter if we deserialize it or not. So we need a string representation of the body anyway.

**Comment by davidfowl** [comment]
> Fully understand your point about built-in formatters and buffering, but we have to trace each body completely within the controller, no matter if we deserialize it or not. So we need a string representation of the body anyway.

https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AspNetCoreGuidance.md#avoid-reading-large-request-bodies-or-response-bodies-into-memory

😄

**Comment by alastairs** [comment]
@davidfowl That link provides context as to _why_ not to do this (great! 😃), but not _what to do instead_ (oh 😞). Let's say you need to accept arbitrary file bodies in a request, and send that arbitrary file request body to another system, such as a database or a message queue. How do you avoid reading that request into memory?

**Comment by berndku** [comment]
Sure, this is not a good practice, but requests are about 20-40KB, so no LOH fragmentation issues expected. I recommended to move the body tracing to a different place in the architecture a while ago, but this will take longer and we needed a short term solution.

Since #13834 is about _HttpRequestStreamReader_,  I checked our custom formatter (code below) again, but there we just used _StreamReader_. Could that be the reason, why we don´t see the blocking with the custom formatter ? 

If yes, does that mean we should avoid _HttpRequestStreamReader_ until the issue is fixed ? 

```csharp
 public class RawJsonBodyInputFormatter : InputFormatter
    {
        public RawJsonBodyInputFormatter()
        {
            this.SupportedMediaTypes.Add("application/json");
        }

        public override async Task<InputFormatterResult> ReadRequestBodyAsync(InputFormatterContext context)
        {
            var request = context.HttpContext.Request;
            using (var reader = new StreamReader(request.Body))
            {
                var content = await reader.ReadToEndAsync();
                return await InputFormatterResult.SuccessAsync(content);
            }
        }

        protected override bool CanReadType(Type type)
        {
            return type == typeof(string);
        }
    }
```

**Comment by vankampenp** [comment]
mm, sorry if this is not related, but I get this output from the blockingmonitor when IIS is restarted. It occurs multiple times. I use the asp-append-version taghelper. After temporarily removing that, the blocking call is no longer reported.

2019-09-10 15:06:25.0604|35.230.16.48||WARN |9.9.19.0|39 |Base|BlockingMonitor.BlockingStart|Blocking method has been invoked and blocked, this can lead to threadpool starvation.
   at System.IO.FileStream.Read(Byte[] array, Int32 offset, Int32 count)
   at System.Security.Cryptography.HashAlgorithm.ComputeHash(Stream inputStream)
   at Microsoft.AspNetCore.Mvc.Razor.Infrastructure.DefaultFileVersionProvider.GetHashForFile(IFileInfo fileInfo)
   at Microsoft.AspNetCore.Mvc.Razor.Infrastructure.DefaultFileVersionProvider.AddFileVersionToPath(PathString requestPathBase, String path)
   at Microsoft.AspNetCore.Mvc.TagHelpers.ScriptTagHelper.Process(TagHelperContext context, TagHelperOutput output)
   at Microsoft.AspNetCore.Razor.TagHelpers.TagHelper.ProcessAsync(TagHelperContext context, TagHelperOutput output)
   at Microsoft.AspNetCore.Razor.Runtime.TagHelpers.TagHelperRunner.RunAsync(TagHelperExecutionContext executionContext)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Razor.Runtime.TagHelpers.TagHelperRunner.RunAsync(TagHelperExecutionContext executionContext)
   at AspNetCore.Views_Shared__Layout.<>c__DisplayClass14_0.<<ExecuteAsync>b__1>d.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at AspNetCore.Views_Shared__Layout.<>c__DisplayClass14_0.<ExecuteAsync>b__1()
   at Microsoft.AspNetCore.Razor.Runtime.TagHelpers.TagHelperExecutionContext.SetOutputContentAsync()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Razor.Runtime.TagHelpers.TagHelperExecutionContext.SetOutputContentAsync()
   at AspNetCore.Views_Shared__Layout.ExecuteAsync()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at AspNetCore.Views_Shared__Layout.ExecuteAsync()
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageCoreAsync(IRazorPage page, ViewContext context)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderPageAsync(IRazorPage page, ViewContext context, Boolean invokeViewStarts)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderLayoutAsync(ViewContext context, ViewBufferTextWriter bodyWriter)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderLayoutAsync(ViewContext context, ViewBufferTextWriter bodyWriter)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Mvc.Razor.RazorView.RenderAsync(ViewContext context)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ViewContext viewContext, String contentType, Nullable`1 statusCode)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ViewContext viewContext, String contentType, Nullable`1 statusCode)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ActionContext actionContext, IView view, ViewDataDictionary viewData, ITempDataDictionary tempData, String contentType, Nullable`1 statusCode)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewExecutor.ExecuteAsync(ActionContext actionContext, IView view, ViewDataDictionary viewData, ITempDataDictionary tempData, String contentType, Nullable`1 statusCode)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewResultExecutor.ExecuteAsync(ActionContext context, ViewResult result)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Mvc.ViewFeatures.ViewResultExecutor.ExecuteAsync(ActionContext context, ViewResult result)
   at Microsoft.AspNetCore.Mvc.ViewResult.ExecuteResultAsync(ActionContext context)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Mvc.ViewResult.ExecuteResultAsync(ActionContext context)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeResultAsync(IActionResult result)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeResultAsync(IActionResult result)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResultFilterAsync[TFilter,TFilterAsync]()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResultFilterAsync[TFilter,TFilterAsync]()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeResultFilters()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeResultFilters()
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Internal.ResourceInvoker.InvokeNextResourceFilter()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext()
   at System.Runtime.CompilerServices.TaskAwaiter.<>c.<OutputWaitEtwEvents>b__12_0(Action innerContinuation, Task innerTask)
   at System.Threading.QueueUserWorkItemCallback`1.<>c.<.cctor>b__6_0(Object state)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.QueueUserWorkItemCallback`1.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()

**Comment by davidfowl** [comment]
> @davidfowl That link provides context as to why not to do this (great! 😃), but not what to do instead (oh 😞). Let's say you need to accept arbitrary file bodies in a request, and send that arbitrary file request body to another system, such as a database or a message queue. How do you avoid reading that request into memory?

By chunking it into < 85K buffers then transferring it chunk by chunk. It's one of the tricks that Stream uses to copy data from one Stream to another Stream https://github.com/dotnet/corefx/blob/ee9995f31b684a0c6e5488eceb2500bf0057da89/src/Common/src/CoreLib/System/IO/Stream.cs#L31-L34. 

That combined with pooling those buffers would result in better performance overall as the large object heap wouldn't be churning aggressively. Also there's a potential security risk in exposing an endpoint that reads everything into memory and transfers it to another system. Any place you can take arbitrary client input is a potential place where clients can potentially cause a denial of service if you're not careful.

**Comment by davidfowl** [comment]
@berndku

> Since #13834 is about HttpRequestStreamReader, I checked our custom formatter (code below) again, but there we just used StreamReader. Could that be the reason, why we don´t see the blocking with the custom formatter ?

Yes, `StreamReader` overrides the right methods.

**Comment by HengzheLi** [comment]
`.UseKestrel(opt =>
{
opt.Limits.MinRequestBodyDataRate = null;
})`
Could this work?

**Comment by VitaliiIsaenko** [comment]
Have the same problem when two services communicate with each other (within localhost). @HengzheLi that didn't help me...

**Comment by AlekseyKrainev** [comment]
Having the same problem

**Comment by clement911** [comment]
Hi all,
We're starting a new project that involves again saving webhooks to the database for further processing.
We expect small payloads and but a high number of requests.

Is the following code optimal?

```
[HttpPost]
        public async Task<OkResult> Webhook()
        {
            var body = await new HttpRequestStreamReader(Request.Body, Encoding.UTF8).ReadToEndAsync();
            //save to Db asynchronously?
            return Ok();
        }
```

**Comment by nateous** [comment]
I am getting this error too, but I think my issue is here:
`blob.UploadFromByteArrayAsync(fileByteArray, 0, fileByteArray.Length).Wait();`
vs
`await blob.UploadFromByteArrayAsync(fileByteArray, 0, fileByteArray.Length);`

using `Wait()` will cause the upload to block the thread, whereas using `await` won't.

**Comment by nateous** [comment]
@clement911 that is a good way to process webhooks, and it would be great if your processing is then done by an event queue instead of the web server.

My issue is receiving a url, downloading it, then uploading it! Stupid to do all of that in the web request. So fixing the `Wait()` issue will be moot once I rework this code to just verify that the image url is good, then dump the actual db saves and image uploads to an event processor.

But thank you for keeping up with this issue!

**Comment by highfield** [comment]
FYI, since the time of my [post here](https://github.com/aspnet/AspNetCore/issues/4707#issuecomment-450334499), I started a new project with Asp Net core 2.2 (previous was 2.1). Furthermore, the app is now running on Ubuntu 18 (previous was on 14).

No one failure since that time.

**Comment by clement911** [comment]
I'm still wondering what's the best way to read the whole request body asynchronously.

We were doing `await new HttpRequestStreamReader(Request.Body, Encoding.UTF8).ReadToEndAsync();`
but after upgrading to net core 3.1 we get this issue: https://github.com/aspnet/AspNetCore/issues/13834

Since the fix is planned for .net 5.0, what is the right way for .net 3.1?

**Comment by galaktipus** [comment]
Any commit for the issue?

**Comment by trevors20** [comment]
Since this issue has been marked as a medium security vulnerability by WhiteSource, can anyone tell me a quick summary of what the exploit is here?  This is a VERY long thread going back to 2018.  Thank you.

**Comment by Jessicaward** [comment]
I'm having the same problem

**Comment by Bit-Shifts** [comment]
Having the same issue as well.

**Comment by highfield** [comment]
For those who are still experiencing the issue, did you read my post? https://github.com/dotnet/aspnetcore/issues/4707#issuecomment-557538742

As today, I had ***NO*** issues since over one year on two different servers.

**Comment by mohitvaid** [comment]
Any update on this issue, I am also facing the same exception. The request is pretty small but kestrel throwing 408 with message request arriving too slow.

` "message": "Reading the request body timed out due to data arriving too slowly. See MinRequestBodyDataRate.", "exception": {"Type":"Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException", "StatusCode":408, "TargetSite":"Void Throw(Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.RequestRejectionReason)", "StackTrace":"   at Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException.Throw(RequestRejectionReason reason)\r\n   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.Http1ContentLengthMessageBody.ReadAsyncInternal(CancellationToken cancellationToken)\r\n   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)\r\n   at Microsoft.AspNetCore.WebUtilities.FileBufferingReadStream.ReadAsync(Byte[] buffer, Int32 offset, Int32 count, CancellationToken cancellationToken)\r\n   at Microsoft.AspNetCore.WebUtilities.StreamHelperExtensions.DrainAsync(Stream stream, ArrayPool`1 bytePool, Nullable`1 limit, CancellationToken cancellationToken)\r\n   at Microsoft.AspNetCore.Mvc.Formatters.NewtonsoftJsonInputFormatter.ReadRequestBodyAsync(InputFormatterContext context, Encoding encoding)\r\n   at Microsoft.AspNetCore.Mvc.ModelBinding.Binders.BodyModelBinder.BindModelAsync(ModelBindingContext bindingContext)\r\n   at Microsoft.AspNetCore.Mvc.ModelBinding.ParameterBinder.BindModelAsync(ActionContext actionContext, IModelBinder modelBinder, IValueProvider valueProvider, ParameterDescriptor parameter, ModelMetadata metadata, Object value)\r\n   at Microsoft.AspNetCore.Mvc.Controllers.ControllerBinderDelegateProvider.<>c__DisplayClass0_0.<<CreateBinderDelegate>g__Bind|0>d.MoveNext()\r\n--- End of stack trace from previous location where exception was thrown ---\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextExceptionFilterAsync>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ExceptionContextSealed context)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|24_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|19_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)\r\n   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)\r\n   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)\r\n   `

**Comment by xcl-net** [comment]
[HttpPost]
        [ProducesResponseType(typeof(BaseResponse<bool>), 1)]
        public async Task<BaseResponse> SyncMsg()
        {

            string data = string.Empty;
            try
            {
                using (var sr = new StreamReader( Request.Body))
                {
                    data = await sr.ReadToEndAsync();

                    //todo: lenth more then 1024, Decompress
                    data = StringHandleHelper.Decompress(data);

                    var url = ConfigurationManager.GetValue("CallBackUrl");

#pragma warning disable CS4014 // 由于此调用不会等待，因此在调用完成前将继续执行当前方法
                    Task.Factory.StartNew((obj) =>
                    {
                         _messageService.SyncMsg((string)obj, url);
                    }, data);
#pragma warning restore CS4014 // 由于此调用不会等待，因此在调用完成前将继续执行当前方法


                    var j_data = JToken.Parse(data);
                    var type = j_data["type"].ToObject<PcWxMsgType>();
                    var wxid = j_data["wxid"].ToObject<string>();

                    return Response2WeChatRobot(wxid, type);
                }
            }
            catch (Exception ex)
            {
                return new BaseResponse() { Code = -2, Msg = "站点:" + ex.Message };
            }
        }
I 


I've seen all the discussions. Is there an effective solution now

**Comment by mohitvaid** [comment]
Is this going to get fixed in the upcoming release?

**Comment by kenyeung128** [comment]
> Hi all,
> We're starting a new project that involves again saving webhooks to the database for further processing.
> We expect small payloads and but a high number of requests.
> 
> Is the following code optimal?
> 
> ```
> [HttpPost]
>         public async Task<OkResult> Webhook()
>         {
>             var body = await new HttpRequestStreamReader(Request.Body, Encoding.UTF8).ReadToEndAsync();
>             //save to Db asynchronously?
>             return Ok();
>         }
> ```

@clement911 if moving to DB processing first would solve the ReadToEndAsync issue?

or else you tried using structured class instead of reading the object from Stream instead?

**Comment by Gopichandar** [comment]
Ok, I'm started reading this thread an hour ago and Im hear at the end yet to get solution.   What is the solution now for .NEt 3.1, Im having same issue

**Comment by davidfowl** [comment]
It's because the issue has turned into lots of different issues 😄. There's no repro for this issue and it depends on the environment and client speed (at least the original issue does). We know there's an issue somewhere but it's not clean how we make progress with the information provided here.

**Comment by kenyeung128** [comment]
@Gopichandar I came across this and solved the issue as follow

- refractor most methods using await + async Task method() pattern

 seems the root cause in our case caused by all CPU being used for waiting task to finish, and this causes the exception with BadHttpRequestException: Reading the request body timed out due to data arriving too slowly

**Comment by rasmus-s** [comment]
@davidfowl  I think I was able to reproduce this issue over and over when I debugged my own code. I had made a small movie and Wireshark captures. But cannot share them here in this public github repo, what other more private channels can I reach out to you and the MS team?  

The longer story:
Was making a few million PUT calls to our servers and around 0,08% of them failed also after multiple retries attempts – all had the same message on the server side: `“Reading the request body timed out due to data arriving too slowly. See MinRequestBodyDataRate”`. For sure the server was not idle during the many calls. But the day after the server was almost idle, and I was able to take the failed calls and re-execute them as a single request at the time. To my big surprise was it consistent that the payload failed and I still see same error message in our server logs. 
Had also tried to increase the timeout, to give the server time for finding a free thread, but still same errors just after 10 sec, instead of the default 5 sec:
`options.Limits.MinRequestBodyDataRate = new MinDataRate(120, TimeSpan.FromSeconds(10))`

After hours of investigation I found out that there was two thing I could change to make the call succeed:
1)	Disconnect Ciso VPN.
The VPN splits the payload into smaller TCP packages (here is it 5 packages). It’s always the last package that need to be re-transmitted. See image, the “.135” is my pc and “.57” is server. (when im disconnected from VPN it will send the payload in single TCP package). Was also connected to VPN when made the few millions calls.
![image](https://user-images.githubusercontent.com/8188529/96259512-01015700-0fbe-11eb-9bb2-73b9807077e7.png)


2)	Change the payload size with 1 byte – either add or remove one character to the json payload (off course on a field that is not analyzed on server side)
The payload is still splited into 5 TCP packages - due to connected to VPN.

The payload is ~6Kb.
Its not only from my pc with VPN that we see this error, we see it from our end customers requests round the world without VPN. But this investigation its the closest I had been to debug, analyze and understand this error message, we see in our server logs. 

I don’t think it’s a thread pool issue on the server side, due I am able to reproduce it over and over and a 1 byte change in the payload size can affect the result. And the other day the server handled way more requests.

I had seen  error message `“Reading the request body timed out due to data arriving too slowly. See MinRequestBodyDataRate”` on both request to routes with simple aspnet core model Binding [FromBody], and routes that read body it self `await Request.Body.ReadAsBytesAsync()`.

So the question is, what is it in aspnetcore that can cause this issue ? and how do I share more data with you guys?

**Comment by norouly** [comment]
It seems like I found a workaround until this gets fixed. 
I had the same issues as @rasmus-s . It kept working until the TCP packages got split into separate ones, which (for me) was exactly at 1398 bytes. Adding the overhead, that's oddly close to the ~1500 bytes of an ethernet frame. 
After doing some research I found a post where someone suggested to check the MTU size, which I did with `ping ip -f -l xxx`. Sure enough, `-l 1473` returned the message `Packet needs to be fragmented but DF set`, so 1472 seems to be the limit where packages won't be split and where exception won't occur. 
To compare my result with something else, I hopped on a system where it worked fine and ran that command again - this time no `Packet needs to be fragmented but DF set` returned. I even used `-l 65500` which is the maximum - no message. 
It actually does seem like the MTU is the problem here.

I'll change the MTU of the first system where it doesn't work and keep this updated.

**Comment by kevbry** [comment]
@davidfowl any chance to reopen this? Looks like rasmus-s found a repro

**Comment by davidfowl** [comment]
@halter73 check out the latest comments

**Comment by Leonardo-Ferreira** [comment]
I think its worth mentioning that sending/reading data slowly is a vector of [DDoS attack](https://www.netscout.com/what-is-ddos/slow-read-attacks)

**Comment by halter73** [comment]
> I think its worth mentioning that sending/reading data slowly is a vector of DDoS attack

Correct. It's a [slowloris attack](https://en.wikipedia.org/wiki/Slowloris_(computer_security)) and that's exactly what this min rate timeout and resulting exception is designed to mitigate.

**Comment by BrandonPotter** [comment]
I don’t think anyone is arguing that it’s not a sane default... but for applications that need to potentially receive data slowly, looking for a reliable way to disable this.

**Comment by davidfowl** [comment]
There is, you can set the minimum data rate (the error itself has a pointer to the thing you need to change) https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.server.kestrel.core.kestrelserverlimits.minrequestbodydatarate?view=aspnetcore-5.0
